File: 0148-sort-list/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode Mid(ListNode head){
        ListNode slow=head;
        ListNode fast=head;
        while(fast.next!=null && fast.next.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        return slow;
    }
    private ListNode mySort(ListNode firstHalf,ListNode secondHalf){
        if(firstHalf==null || secondHalf==null){
            return (firstHalf==null)?secondHalf:firstHalf;
        }
        ListNode myNode=new ListNode(7);
        ListNode curr=myNode;
        while(firstHalf!=null && secondHalf!=null){
            if(firstHalf.val<secondHalf.val){
                curr.next=firstHalf;
                firstHalf=firstHalf.next;
            }else{
                curr.next=secondHalf;
                secondHalf=secondHalf.next;
            }
            curr=curr.next;
        }
        if(firstHalf!=null || secondHalf!=null){
            curr.next=(firstHalf!=null)?firstHalf:secondHalf;
        }
        return myNode.next;
    }
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null) return head;
        ListNode mid=Mid(head);
        ListNode newHead=mid.next;
        mid.next=null;
        ListNode firstHalf=sortList(head);
        ListNode secondHalf=sortList(newHead);
        return mySort(firstHalf,secondHalf);
    }
}

File: 0148-sort-list/0148-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* MID(ListNode* head){
        ListNode* slow=head;
        ListNode* fast=slow;
        while(fast->next!=NULL && fast->next->next!=NULL){
            slow=slow->next;
            fast=fast->next->next;
        }
        return slow;
    }
    ListNode* myFunction(ListNode* left,ListNode* right){
        if(left==NULL || right==NULL) return (left==NULL)?right:left;
        ListNode *newNode=new ListNode(7);
        ListNode *curr=newNode;
        while(left!=NULL && right!=NULL){
            if(left->val<right->val){
                curr->next=left;
                left=left->next;
            }else{
                curr->next=right;
                right=right->next;
            }
            curr=curr->next;
        }
        curr->next=(left==NULL)?right:left;
        return newNode->next;
    }
public:
    ListNode* sortList(ListNode* head) {
        if(head==NULL || head->next==NULL) return head;
        ListNode* midNode=MID(head);
        ListNode* newNode=midNode->next;
        midNode->next=NULL;
        ListNode *left_sub_list=sortList(head);
        ListNode *right_sub_list=sortList(newNode);
        return myFunction(left_sub_list,right_sub_list);
    }
};